contracts/src/v0.1/L1Endpoint.sol

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./VRFConsumerBase.sol";
+import "./interfaces/IVRFCoordinator.sol";
+
+contract L1Endpoint is Ownable, VRFConsumerBase {
+    IVRFCoordinator COORDINATOR;
+    mapping(address => uint256) private _balance;
+    mapping(address => bool) private _oracles;
+    mapping(uint256 => address) private _requestIdRequester;
+    uint256 private _fee;
+
+    error InsufficientBalance();
+    error OnlyOracle();
+
+    event FeeUpdated(uint256 oldFee, uint256 newFee);
+    event OracleAdded(address oracle);
+    event OracleRemoved(address oracle);
+
+    event RandomWordRequested(uint256 requestId, address requester);
+    event RandomWordFulfilled(uint256 requestId, address requester, uint256[] randomWords);
+
+    constructor(address coordinator) VRFConsumerBase(coordinator) {
+        COORDINATOR = IVRFCoordinator(coordinator);
+    }
+
+    receive() external payable {
+        _balance[msg.sender] += msg.value;
+    }
+
+    function setFee(uint256 newFee) public onlyOwner {
+        uint256 cFee = _fee;
+        _fee = newFee;
+        emit FeeUpdated(cFee, newFee);
+    }
+
+    function addOracle(address oracle) public onlyOwner {
+        _oracles[oracle] = true;
+        emit OracleAdded(oracle);
+    }
+
+    function removeOracle(address oracle) public onlyOwner {
+        delete _oracles[oracle];
+        emit OracleRemoved(oracle);
+    }
+
+    function requestRandomWordsDirectPayment(
+        bytes32 keyHash,
+        uint32 callbackGasLimit,
+        uint32 numWords,
+        address payer,
+        address requester
+    ) public returns (uint256 requestId) {
+        if (!_oracles[msg.sender]) {
+            revert OnlyOracle();
+        }
+        if (_balance[payer] < _fee) {
+            revert InsufficientBalance();
+        }
+        _balance[payer] = _balance[payer] - _fee;
+
+        uint256 id = COORDINATOR.requestRandomWords{value: _fee}(
+            keyHash,
+            callbackGasLimit,
+            numWords,
+            msg.sender
+        );
+        _requestIdRequester[id] = requester;
+        emit RandomWordRequested(id, requester);
+        return id;
+    }
+
+    function fulfillRandomWords(
```

Comment:
Maybe I miss something, but when is this function going to be called?

Reply:
It's gonna be called by our offchain oracles on L1. In this scenario, L1Enpoint serves as a consumer

Reply:
So this function implementation is not finished yet, right?

Reply:
It is finished at this point.
the main function of it is just emit the event with the result, RequestID and the address of the requester on L2. With those parameters, our L2 offchain oracles can fullfil for the consumer on L2

Reply:
Aha! Now, I understand. Thank you!